### ISSUE TITLE
Integrating Electron Builder with React resulting in bigger app.

### ISSUE THREAD

**Issue by primalanomaly** [electron-userland/electron-builder#3325]
Hi, I'm trying to integrate Electron Builder with React (create-react-app).

Currently my react code and my electron code are separate projects, and work fine, but I'd rather integrate them properly to speed up the build process.

At present, I build the React app, and use a postbuild script to copy the built files to a directory in my electron app. I then build the electron app, pointing the url to this directory.

When I tried integrating, I assumed I should set up my scripts to build the React app first and then build the Electron app, again pointing to the React directory. However, the compiled app is twice the size of before (280mb vs 150mb), and takes longer for Electron Builder to package.

Is Electron-Builder doing some automatic magic with React and resulting in 2 React builds, or something like that? Or is it compiling all of the NPM modules for React into the Electron project, or something similar? I can't really work out why this is happening, but it seems like my understanding of the build steps is obviously wrong somewhere. Thanks.

**Comment by odiseo42** [comment]
> At present, I build the React app, and use a postbuild script to copy the built files to a directory in my electron app. I then build the electron app, pointing the url to this directory.

Is the size incrementing the more builds you create? Probably you're just replicating files on each one of your React builds.

**Comment by nurpax** [comment]
I managed to make a CRA 2.1 based Electron app (https://github.com/nurpax/petmate) and the build process works fine.  However, I also experience too much bundling of node_modules in the Electron package builds.  I profiled this by extracting the generated app.asar and noticed that a lot of unnecessary stuff goes into my app.asar (the archived node_modules is roughly 42MB).  Most likely almost all of this is redundant as anything that goes into the renderer part is built by CRA's build target which bundles all the npm packages into a few minized .js files (less than 2MB total).

Electron-builder packs packages you list under "dependencies" into the app.asar.  I did a quick test that moving "all the deps that are needed for the renderer build" (e.g., CRA build) to devDependencies slims down the compressed app.asar size by some megabytes.  When I moved all my dependencies to devDependencies, the node_modules directory disappeared (as expected) from app.asar while my packaged app still worked.  The size delta in my case was not super huge (some megabytes) but I guess it all depends on how many deps you're pulling into your project.

I guess it's not really "wrong" to use devDependencies for listing the libraries your renderer js bundle needs.  After all, it's just the compiled bundle that really matters.  But someone with more expertise on these tools would know better.
